### PR TITLE
DOC: Improve trapz docstring.

### LINKED ISSUE #16136: trapz() should raise a warning if x is not sorted.
trapz computes wrong result if x is not sorted. Raising a warning if x is out of order will be quite helpful.

### PR DESCRIPTION
**Pull request by brpy**
Closes #16136

DOC: Added trapz function documentation changes wrto. issue #16136 

Added that x points are assumed to be in sorted order and unique.

### CODE REVIEW

**Review by eric-wieser — CHANGES_REQUESTED** [review]
Thanks for the contribution, but as I just commented in the tracking issue, these changes are not correct. `trapz(x, y)` integrates a parametric curve in the order of your samples. The fact that such a curve may go back on itself is irrelevant, trapz still does the mathetmatically correct thing.

File: numpy/lib/function_base.py [hunk 1]
```diff
@@ -4054,6 +4054,10 @@ def trapz(y, x=None, dx=1.0, axis=-1):
     4.0
     >>> np.trapz([1,2,3], x=[4,6,8])
     8.0
+    >>> np.trapz([0,1,0], x=[0,1,2])
+    1.0
+    >>> np.trapz([0,0,1], x=[0,2,1])
+    -0.5
```
Comment (eric-wieser) [inline numpy/lib/function_base.py:4060]: Perhaps more insightful would be to repeat the example above with the x coordinates in reverse order, and to deliberately integrate a closed curve:

```suggestion
    # integrate in reverse
    >>> np.trapz([1,2,3], x=[8,6,4])  
    -8.0

    # Find the area of a circle
    >>> theta = np.linspace(0, 2 * np.pi, num=1000, include_endpoint=True)
    >>> np.trapz(np.cos(theta), x=np.sin(theta))
    3.141571941375841
```

File: numpy/lib/function_base.py [hunk 2]
```diff
@@ -4008,7 +4008,7 @@ def trapz(y, x=None, dx=1.0, axis=-1):
     """
     Integrate along the given axis using the composite trapezoidal rule.
 
-    Integrate `y` (`x`) along given axis.
+    Integrate `y` (`x`) along given axis, in the order of values passed.
```
Comment (eric-wieser) [inline numpy/lib/function_base.py:4011]: Wording this is tricky, maybe worth mentioning this only applies if `x` is given.
```suggestion
    Integrate `y` (`x`) along given axis. If `x` is provided, the integration happens in sequence along its elements - they are not sorted.
```

Could also embed some latex here (if you replace `"""` with `r"""` above) with

```suggestion
    Along each 1d slice on the given axis, compute :math:`\int y(x) dx`.
    When `x` is specified, this integrates along the parametric curve, computing :math:`\int_t y(t) dt = \int_t y(t) \left.\frac{dx}{dt}\right|_{x=x(t)} dt`.
```

Let's see what other reviewers have to say.

File: numpy/lib/function_base.py (hunk 1 above)
Comment (brpy) [reply numpy/lib/function_base.py]: Nice intuitive examples!

File: numpy/lib/function_base.py [hunk 3]
```diff
@@ -4054,6 +4054,15 @@ def trapz(y, x=None, dx=1.0, axis=-1):
     4.0
     >>> np.trapz([1,2,3], x=[4,6,8])
     8.0
+    # integrate in reverse
+    >>> np.trapz([1,2,3], x=[8,6,4])  
+    -8.0
+
+    # Integrate along a parametric curve.
+    # This finds the area of a circle, noting we repeat the sample which closes the curve.
+    >>> theta = np.linspace(0, 2 * np.pi, num=1000, include_endpoint=True)
```
Comment (eric-wieser) [inline numpy/lib/function_base.py:4063]:
```suggestion
    >>> theta = np.linspace(0, 2 * np.pi, num=1000, endpoint=True)
```
Whoops

**Comment by eric-wieser** [comment]
Could also embed some latex here (if you replace `"""` with `r"""` in the docstring) with

```rst
    Along each 1d slice on the given axis, compute :math:`\int y(x) dx`.
    When `x` is specified, this integrates along the parametric curve, computing :math:`\int_t y(t) dt = \int_t y(t) \left.\frac{dx}{dt}\right|_{x=x(t)} dt`.
```

Let's see what other reviewers have to say.

**Comment by brpy** [comment]
Some build-tests are failing.

**Comment by rossbar** [comment]
@brpy I'm a bit confused about the changes introduced in a4c54d3 - they seem incomplete and at-odds with some of the previous review comments. Did you perhaps push an unfinished commit? 

I'm happy to take a look but am having a hard time trying to disentangle the most recent changes.

File: numpy/lib/function_base.py [hunk 4]
```diff
@@ -4005,11 +4005,14 @@ def _trapz_dispatcher(y, x=None, dx=None, axis=None):
 
 @array_function_dispatch(_trapz_dispatcher)
 def trapz(y, x=None, dx=1.0, axis=-1):
-    """
+    r"""
     Integrate along the given axis using the composite trapezoidal rule.
 
-    Integrate `y` (`x`) along given axis.
-
+    If `x` is provided, the integration happens in sequence along its elements - they are not sorted.
```
Comment (eric-wieser) [inline numpy/lib/function_base.py:4011]: This comment feels out of place here - maybe move it down to the parameters section?

**Comment by brpy** [comment]
@eric-wieser  From your previous suggestions, 

Integrate `y` (`x`) along each 1d slice on the given axis, compute :math:`\int y(x) dx`.
    When `x` is specified, this integrates along the parametric curve, computing :math:`\int_t y(t) dt = \int_t y(t) \left.\frac{dx}{dt}\right|_{x=x(t)} dt`.
   
seems to be at odds with previous comments as @rossbar rightly points out. If `x` is provided integration happens in the order of `x` provided but not on the parametric curve.

Can you guys please look into it and suggest some final changes? 
I will make changes accordingly.

**Comment by rossbar** [comment]
FWIW I don't disagree with @eric-wieser 's suggestion - my main issue is that some of the changes in a4c54d3 seem to be unprompted (and not as clear to me). For example, I'm against replacing

> Integrate `y` (`x`) along given axis.

with

> Integrate `y` (`x`) along each 1d slice on the given axis, compute :math:`\int y(x) dx`.

I didn't see that this was suggested anywhere in the previous discussion, though maybe I missed a resolved conversation somehwere. That's part of the reason why I was unsure about whether all of the changes in a4c54d3 were intentional. This could very well just be my own inability to follow the discussions on GH - please let me know if there is something I am missing.

**Comment by eric-wieser** [comment]
@rossbar, did you see https://github.com/numpy/numpy/pull/16187#issuecomment-625830797?

**Comment by eric-wieser** [comment]
> seems to be at odds with previous comments as @rossbar rightly points out. If `x` is provided integration happens in the order of `x` provided but not on the parametric curve.

I don't really understand what you mean by that. To reiterate and slightly improve my claim from an earlier comment:

> When `x` is specified, this integrates along the parametric curve `x(t), y(t)`, computing :math:`\int_t y(t) dt = \int_t y(t) \left.\frac{dx}{dt}\right|_{x=x(t)} dt`, where `t` is the index in the array.

**Comment by rossbar** [comment]
> @rossbar, did you see #16187 (comment)?

Nope, I had missed it - thanks for pointing me to it!

**Comment by rossbar** [comment]
This PR was stuck on some CI failures, so I went ahead and fixed the outstanding rST formatting issues to try to get this over the hump. @eric-wieser maybe you could take another look to see if your comments are all sufficiently addressed.

**Comment by brpy** [comment]
Any changes to make to this PR that would help it get merged?

**Comment by eric-wieser** [comment]
Sorry for dropping the ball on this. I think this is good to go; can you merge in the latest `main` to check it passes the latest CI?

**Comment by brpy** [comment]
@eric-wieser I have merged latest `main` in a new pr  #19041


I have deleted my old fork so made a new one. My apologies.

**Pull request closed without merging**